TypeError bug in python3

captam3rica opened the pull request:
TypeError: a bytes-like object is required, not 'str'

Receive this error when attempting to generate the bootstrap.json file on python 3.8.1 and 3.7.6. However, the current code still works on 2.7.x. Tested the proposed fix on 3.8.1, 3.7.6, and 2.7.16, so it should be backward compatible with older versions of 3.8.

Traceback (most recent call last):
  File "generatejson.py", line 276, in <module>
    main()
  File "generatejson.py", line 249, in main
    (pkgId, pkgVersion) = getpkginfo(filePath)
  File "generatejson.py", line 93, in getpkginfo
    pkgInfoPath = extractpkginfo(filename)
  File "generatejson.py", line 74, in extractpkginfo
    pkgInfoPath = getpkginfopath(filename)
  File "generatejson.py", line 52, in getpkginfopath
    bom = bom.strip().split("\n")
TypeError: a bytes-like object is required, not 'str'

captam3rica commented on generatejson.py:
```diff
@@ -49,15 +49,18 @@ def getpkginfopath(filename):
+    
+    try:
+        # Decode the byte stream and convert to list
+        bom = bom.decode("utf-8").split()
```
Decode the byte stream and convert to list that we can then parse and look for the things that we need.

captam3rica commented on generatejson.py:
```diff
@@ -94,8 +97,10 @@ def getpkginfo(filename):
         pkgRefs = dom.getElementsByTagName('pkg-info')
         for ref in pkgRefs:
-            pkgId = ref.attributes['identifier'].value.encode('UTF-8')
-            pkgVersion = ref.attributes['version'].value.encode('UTF-8')
```
Caused the identifier and version keys in the bootstrap.json file to not be retrieved because JSON could not decode the byte object.

erikng commented:
Is this failing on a specific package? I'm curious as we've been using the current version for some time without issue.

captam3rica commented:
No nothing specific. It has actually failed on packages that I have created with Packages.app or that I have downloaded from the web like JamfConnectVerify or the VMwareWorkspaceONE agent installer package.

At first, I thought that it might just be my environment, but I had someone else try it on their Mac as well and they were getting the same thing on python version 3.7.3. Then, they ran the script with the updates and it went through on both 2.7 and 3.8.

I wonder if there's something else I could try to validate further? I used pyenv to switch back and forth among versions of python. And tested on vanilla installs of macOS.

erikng commented:
I'll have to confirm internally but pretty sure I only tested on 3.8, which is what InstallApplications uses. Across three of my tools I've only tested against 3.8 and I don't really plan on supporting anything lower. The main issue is I don't want to force my python onto others. It may be that I should remove the shebang entirely and have a note that you will need to pass a python3.8 release to it.

captam3rica commented:
I'll give "python3.8 generatejson.py" a shot on my side and report back. Thanks for taking the time to respond 😊

captam3rica commented:
Still ran into the same outcome with python3.8 defined as an interpreter.

Avengers%: python3.8 generatejson.py --base-url https://github.com --output ~/Desktop \
--item \
item-name='JamfConnectLogin' \
item-path='/Users/captam3rica/OneDrive - Insight/Git Repos/macos-bootstrap/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \ 
item-stage='setupassistant' \
item-type='package' \
item-url='https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \
script-do-not-wait=False
Traceback (most recent call last):
  File "generatejson.py", line 276, in <module>
    main()
  File "generatejson.py", line 249, in main
    (pkgId, pkgVersion) = getpkginfo(filePath)
  File "generatejson.py", line 93, in getpkginfo
    pkgInfoPath = extractpkginfo(filename)
  File "generatejson.py", line 74, in extractpkginfo
    pkgInfoPath = getpkginfopath(filename)
  File "generatejson.py", line 52, in getpkginfopath
    bom = bom.strip().split("\n")
TypeError: a bytes-like object is required, not 'str'

If no one else has reported a similar issue then we can go ahead and close this pull request.

erikng commented:
I think I see the issue now. Can you do me a favor and remove the space in your path? The one drive thing.

My guess is the file path is trying to be stripped because of that, and that code wasn't converted to a byte string. 

This could also be why I didn't see it and others haven't reported.

captam3rica commented:
Will do and let you know how it goes. I have also Tried directly from the desktop but will try a few more variations of the path.

captam3rica commented:
With **python 3.8** as the version running the script from a directory in my Desktop.

Avengers%: python3.8 --version                                                           
Python 3.8.0
Avengers%: python3.8 generatejson.py --base-url https://github.com --output ~/Desktop \ 
--item \
item-name='JamfConnectLogin' \
item-path='/Users/captam3rica/Desktop/macos-bootstrap/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \
item-stage='setupassistant' \
item-type='package' \
item-url='https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \
script-do-not-wait=False
Traceback (most recent call last):
  File "generatejson.py", line 266, in <module>
    main()
  File "generatejson.py", line 239, in main
    (pkgId, pkgVersion) = getpkginfo(filePath)
  File "generatejson.py", line 89, in getpkginfo
    pkgInfoPath = extractpkginfo(filename)
  File "generatejson.py", line 72, in extractpkginfo
    pkgInfoPath = getpkginfopath(filename)
  File "generatejson.py", line 50, in getpkginfopath
    bom = bom.strip().split("\n")
TypeError: a bytes-like object is required, not 'str'

With the macOS version of **python2(2.7.16**) as the main version.

Avengers%: pyenv global system
Avengers%: python --version                                                         
Python 2.7.16
Avengers%: python generatejson.py --base-url https://github.com --output ~/Desktop \ 
--item \
item-name='JamfConnectLogin' \
item-path='/Users/captam3rica/Desktop/macos-bootstrap/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \
item-stage='setupassistant' \
item-type='package' \
item-url='https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg' \
script-do-not-wait=False
Json saved to /Users/captam3rica/Desktop/bootstrap.json

erikng commented:
Any chance you could give me this package? I think the solution is reasonable I just want to understand why we aren't seeing this issue.

Thanks,
Erik Gomez
________________________________
From: captam3rica <notifications@github.com>
Sent: Wednesday, February 26, 2020 1:17:08 PM
To: erikng/installapplications <installapplications@noreply.github.com>
Cc: Erik Gomez <e@eriknicolasgomez.com>; Review requested <review_requested@noreply.github.com>
Subject: Re: [erikng/installapplications] TypeError bug in python3 (#61)


With python 3.8 as the version running the script from a directory in my Desktop.

Avengers%: python3.8 --version
Python 3.8.0
Avengers%: python3.8 generatejson.py --base-url https://github.com --output ~/Desktop \
--item
item-name='JamfConnectLogin'
item-path='/Users/captam3rica/Desktop/macos-bootstrap/payload/setupassistant/jamf-connect-login-bootstrap.pkg'
item-stage='setupassistant'
item-type='package'
item-url='https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg'
script-do-not-wait=False
Traceback (most recent call last):
File "generatejson.py", line 266, in
main()
File "generatejson.py", line 239, in main
(pkgId, pkgVersion) = getpkginfo(filePath)
File "generatejson.py", line 89, in getpkginfo
pkgInfoPath = extractpkginfo(filename)
File "generatejson.py", line 72, in extractpkginfo
pkgInfoPath = getpkginfopath(filename)
File "generatejson.py", line 50, in getpkginfopath
bom = bom.strip().split("\n")
TypeError: a bytes-like object is required, not 'str'

With the macOS version of python2(2.7.16) as the main version.

Avengers%: pyenv global system
Avengers%: python --version
Python 2.7.16
Avengers%: python generatejson.py --base-url https://github.com --output ~/Desktop \
--item
item-name='JamfConnectLogin'
item-path='/Users/captam3rica/Desktop/macos-bootstrap/payload/setupassistant/jamf-connect-login-bootstrap.pkg'
item-stage='setupassistant'
item-type='package'
item-url='https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg'
script-do-not-wait=False
Json saved to /Users/captam3rica/Desktop/bootstrap.json

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/erikng/installapplications/pull/61?email_source=notifications&email_token=ABLL6GGFOUFRFVYITS75LN3RE253JA5CNFSM4KYZUYZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENBQLGA#issuecomment-591594904>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABLL6GFRLLTI4W2GO6C7JA3RE253JANCNFSM4KYZUYZA>.

captam3rica commented:
Agreed. Although I seem to get the same outcome no matter the package that I feed into the script.

I think you should be able grab this specific package from here https://raw.githubusercontent.com/captam3rica/macos-bootstrap/master/payload/setupassistant/jamf-connect-login-bootstrap.pkg 

It will try to download from my repo. Let me know if that doesn’t work and I will get to you another way.

captam3rica commented:
Could this be the case ...

In Python 3 by default strings are stored as Unicode. Whereas in python2 strings are stored as ASCII by default.

This would make sense then that the “strip().split()” would need to be decoded because the string return was a Unicode byte value and could not be split because it was expecting a Unicode value and not a string value.

I suppose another way of writing it could be non.strip().split(u”\n”) to distinguish the \n as Unicode. Would need to test this theory though.

erikng commented:
Yep I agree that there is likely a bug, I just don't understand why I haven't seen it in the months leading up to this. Either my tests are bad or something weird is happening. I'll check your package tomorrow.

erikng commented:
I have confirmed the issue and now know why my tests were working. My tests were bad.

Please try https://github.com/erikng/installapplications/commit/daf7a759ec0eff7cdf96546c31c885a9aa1766c0 and see if your issue is now resolved.

erikng commented:
I have confirmed the issue and now know why my tests were working. My tests were bad.

Please try https://github.com/erikng/installapplications/commit/daf7a759ec0eff7cdf96546c31c885a9aa1766c0 and see if your issue is now resolved.

captam3rica commented:
That works!
